diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6d71c9d..cf39550 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,15 +12,23 @@ if ("${CMAKE_BUILD_TYPE}" STREQUAL "")
   set(CMAKE_BUILD_TYPE RELWITHDEBINFO)
 endif()
 set(CMAKE_VERBOSE_MAKEFILE OFF)
-option(BUILD_MARCH_NATIVE "gcc flag -march=native" on)
-if(BUILD_MARCH_NATIVE)
-  message("Build with -march=native")
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=native -O2")
+
+
+if(MICROARCH_LEVEL STREQUAL "4")
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=x86-64-v4 -mtune=cascadelake")
+  set(QCINT_VECTOR_LEVEL "avx512f")
+  set(QCINT_SIMDD 8)
+elseif(MICROARCH_LEVEL STREQUAL "3")
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=x86-64-v3 -mtune=cascadelake")
+  set(QCINT_VECTOR_LEVEL "avx")
+  set(QCINT_SIMDD 4)
 else()
-  # This appears to be the lowest version that compiles
-  message("Build with -march=sse3")
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -msse3 -O2")
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -march=nocona -mtune=haswell")
+  set(QCINT_VECTOR_LEVEL "sse3")
+  set(QCINT_SIMDD 2)
 endif()
+
+
 if ("${CMAKE_C_COMPILER_ID}" STREQUAL "Intel")
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -unroll-aggressive -ipo ")
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fno-math-errno")
@@ -52,23 +60,7 @@ check_function_exists(sqrtl HAVE_SQRTL)
 #set(CMAKE_REQUIRED_LIBRARIES quadmath)
 #check_function_exists(fabsq HAVE_QUADMATH_H)
 
-include(CheckNativeVectorization)
-check_native_vectorization()
-message("sse3 works: ${SSE3_WORKS}")
-message("avx works: ${AVX_WORKS}")
-message("avx512f works: ${AVX512F_WORKS}")
-if(AVX512F_WORKS)
-  set(QCINT_VECTOR_LEVEL "avx512f")
-  set(QCINT_SIMDD 8)
-elseif(AVX_WORKS)
-  set(QCINT_VECTOR_LEVEL "avx")
-  set(QCINT_SIMDD 4)
-elseif(SSE3_WORKS)
-  set(QCINT_VECTOR_LEVEL "sse3")
-  set(QCINT_SIMDD 2)
-else()
-  message(FATAL_ERROR "SSE3 is the minimum supported instruction set")
-endif()
+
 
 message("Will compile with vectorization width: ${QCINT_VECTOR_LEVEL}, SIMDD=${QCINT_SIMDD}")
 
